### PR TITLE
Refactor BuilderBase to account for changes to shell, shebang key in script schema

### PR DESCRIPTION
**Pull request by shahzebsiddiqui**
sync script schema from https://github.com/buildtesters/schemas/blob/master/script/script-v0.0.1.schema.json
modify implementation on shell detection, environment variable, and the
command to execute the test. We can now support passing shell options on
the command line. The shebang can be customized via 'shebang' key if
it's not found it will use the default shebang based on shell type. If
it's python it will resort to path of 'python' wrapper.
When running in dry mode we were getting output header of table of
results that were not suppose to show up.
Add more log details for BuilderBase

### CODE REVIEW

**Comment by shahzebsiddiqui** [comment]
@vsoch i have tested this locally for all the shell types and options

I will just give an example for python when ``examples/script/circle-area.py``. Shown below is the generated script, note that the python interpreter is set to path of 'python' and environment is also working.

```
(buildtest) ssi29@ag-mxg-hulk090> cat .buildtest/python-circle/circle_area.py
#!/u/users/ssi29/.local/share/virtualenvs/buildtest-0NCSLNEz/bin/python
import os

os.putenv("foo", "bar")
import math

radius = 2
area = math.pi * radius * radius
print("Circle Radius ", radius)
print("Area of circle ", area)

```

For the other bash/sh examples i updated the tutorials repo https://github.com/buildtesters/tutorials/commit/19ef0a228521fe5221bb0167ed63a0da9d8ef07b where some of these tests can be run.

File: buildtest/buildsystem/base.py
Comment (vsoch) [inline buildtest/buildsystem/base.py]: Do you want to make self.shell a property instead? I can see this being reused.

File: buildtest/buildsystem/base.py
Comment (vsoch) [inline buildtest/buildsystem/base.py]: This will trigger an error if Python isn’t found and returned as None.

**Comment by shahzebsiddiqui** [comment]
I have attached an example buildtest.log for this run. The system directory contains a few examples of the shell and shebang

```
(buildtest) ssi29@ag-mxg-hulk090> buildtest build -b system/

            Discovered Buildspecs

/u/users/ssi29/.buildtest/site/github.com/buildtesters/tutorials/system/systemd.yml
/u/users/ssi29/.buildtest/site/github.com/buildtesters/tutorials/system/selinux.yml
/u/users/ssi29/.buildtest/site/github.com/buildtesters/tutorials/system/ulimits.yml
/u/users/ssi29/.buildtest/site/github.com/buildtesters/tutorials/system/disk_usage.yml



Buildspec Name                 SubTest                        Status                         Buildspec Path
________________________________________________________________________________________________________________________
systemd                        systemd_default_target         PASS                           /u/users/ssi29/.buildtest/site/github.com/buildtesters/tutorials/system/systemd.yml
selinux                        selinux_disable                PASS                           /u/users/ssi29/.buildtest/site/github.com/buildtesters/tutorials/system/selinux.yml
ulimits                        ulimit_filelock                PASS                           /u/users/ssi29/.buildtest/site/github.com/buildtesters/tutorials/system/ulimits.yml
ulimits                        ulimit_cputime                 PASS                           /u/users/ssi29/.buildtest/site/github.com/buildtesters/tutorials/system/ulimits.yml
ulimits                        ulimit_stacksize               FAIL                           /u/users/ssi29/.buildtest/site/github.com/buildtesters/tutorials/system/ulimits.yml
disk_usage                     root_disk_usage                PASS                           /u/users/ssi29/.buildtest/site/github.com/buildtesters/tutorials/system/disk_usage.yml


============================================================
                        Test summary
============================================================
Executed 6 tests
Passed Tests: 5/6 Percentage: 83.333%
Failed Tests: 1/6 Percentage: 16.667%

```
[buildtest.log](https://github.com/buildtesters/buildtest/files/4547521/buildtest.log)

File: buildtest/buildsystem/base.py
Comment (vsoch) [reply buildtest/buildsystem/base.py]: Wouldn’t there be a more elegant way to do this instead of hard coding python?

File: buildtest/buildsystem/base.py
Comment (vsoch) [inline buildtest/buildsystem/base.py]: This would use the self.shell property.

**Review by vsoch — CHANGES_REQUESTED** [review]

File: buildtest/buildsystem/base.py
Comment (shahzebsiddiqui) [reply buildtest/buildsystem/base.py]: Yea if python is not found we can raise an error. Also ``shell: python`` adapts to both python2 or python3 but it's not clear to user when they write in ``run`` section and test is not portable since it depends on user python environment. So if it works for me it may not necessarily work for you.

Right now we don't have ability to pass system python let's say ``shell: /usr/bin/python`` or ``shell: /usr/bin/python3`` similarly.

I particularly don't like having ``env`` support when shell is python, it causes some confusion to user when it sees buildtest modify the python script. I was thinking of removing this support but didnt for time being

File: buildtest/buildsystem/base.py
Comment (shahzebsiddiqui) [reply buildtest/buildsystem/base.py]: Can you describe with example how you plan to reuse the shell decorator. I have been getting up to speed on decorators but haven't found a use-case. 

I am thinking you want to invoke BuilderBase class directly if i understand. I think changing shell from ``bash`` --> ``sh`` or ``sh`` --> ``bash`` would be okay for most part, but there are some minor differences between the two shells that can cause an error depending on how ``run`` section is setup. I was able to find this link https://www.gnu.org/software/bash/manual/html_node/Major-Differences-From-The-Bourne-Shell.html outlining some of the differences between ``bash`` and ``sh`` 

I don't see a usecase where we can allow a shell to change from ``bash`` or ``sh`` --> ``python`` since that would cause run section to break pretty dramatically.

File: buildtest/buildsystem/base.py
Comment (shahzebsiddiqui) [reply buildtest/buildsystem/base.py]: I don't think it's possible for ``python`` to not be found when running buildtest. It would cause an error running buildtest. In the ``master`` branch we had the **buildtest** https://github.com/buildtesters/buildtest/blob/master/bin/buildtest as a shell script that checks for the python client as a requirement. That has been removed in ``devel`` but i can see it being useful having these checks at start of buildtest.

File: buildtest/buildsystem/base.py
Comment (vsoch) [reply buildtest/buildsystem/base.py]: All the times in the code we do some action like "self.recipe.get("shell") we would want to just reference self.shell. And I was thinking this morning, since there are now a bunch of different parameters around a shell, would it make sense to have it be a class? E.g.,

```Python
self.shell = Shell(self.recipe.get('shell', '/bin/bash'))
self.shell.opts
self.shell.shebang
```
etc.

File: buildtest/buildsystem/base.py
Comment (vsoch) [reply buildtest/buildsystem/base.py]: We only have python3 on sherlock, which is what made me think of this. I think (given that there can be multiple shells allowed) instead of hard coding them in specifically, it should be done with a single list, the schema validates, but then the code extracts (whatever the valid thing is). That's a much more elegant implementation than hard coding python into the code (and all the other future cases).

File: buildtest/buildsystem/base.py
Comment (shahzebsiddiqui) [reply buildtest/buildsystem/base.py]: I have been thinking about the ``shell`` key this is specific to ``LocalExecutor`` behavior since we are running test locally. This wont be relevant when running test with ``SlurmExector``. It's true we still need to detect the shell for writing the test, get test extension,  but execution of test is depended on the executor type. We also have ``executor`` key https://github.com/buildtesters/buildtest/blob/devel/buildtest/buildsystem/schemas/script/script-v0.0.1.schema.json#L39  which we have not implemented during the build stage. Right now everything defaults to LocalExecutor but again this is out of scope for this PR but worth pointing out.

File: buildtest/buildsystem/base.py
Comment (shahzebsiddiqui) [reply buildtest/buildsystem/base.py]: Are you referring to modifying the schema line https://github.com/buildtesters/schemas/blob/master/script/script-v0.0.1.schema.json#L44 

Do you want add python3 and system python (/bin/python, /bin/python3) into the pattern

```
"pattern": "^(/bin/bash|/bin/sh|sh|bash|python|python3|/bin/python|/bin/python3).*"
```

The python interpreter will be whatever python wrapper was used when user is invoking buildtest so it's not possible to not have python in $PATH. I think having the distinction between python3 and python2 would be useful so user knows how to write their script in ``run`` section.

Comment (vsoch) [reply buildtest/buildsystem/base.py]: I think I've mentioned this before - in that shell is an executable, I don't think we should limit the user to some known set. I should be able to run bash, sh (or a specific location, /bin/sh) or python3, or python4 when it comes out. We shouldn't need to pre-define the executables that the use can specify, only look for them with shutil.which() and exit with a clear error if they chose something not findable.

Comment (shahzebsiddiqui) [reply buildtest/buildsystem/base.py]: If i understand you like to run ``shutil.which()`` on ``self.shell_name`` and if its ``None`` raise an error.  This would be much earlier in the code, but note that we have ``BuildTestCommand`` raise this error at https://github.com/buildtesters/buildtest/blob/devel/buildtest/utils/command.py#L46 which would eventually catch this exception.

Comment (vsoch) [reply buildtest/buildsystem/base.py]: Yep.

Comment (shahzebsiddiqui) [reply buildtest/buildsystem/base.py]: See https://github.com/buildtesters/buildtest/pull/290/commits/922756693cf0320010121c16e3deb243b2c756ba

File: buildtest/buildsystem/base.py
Comment (shahzebsiddiqui) [reply buildtest/buildsystem/base.py]: I dont feel comfortable making this change, i think this relates to discussion  https://github.com/buildtesters/buildtest/pull/290#discussion_r417504156

Comment (vsoch) [reply buildtest/buildsystem/base.py]: Yep happy to review this again in another PR.

**Comment by vsoch** [comment]
Since this PR is directly relevant to factoring in changes to shell, I think the previous discussion about a class for the Shell (and a much simpler interface here) is scoped within this PR, and I'd like to see those changes here to clean up the currently long logic to just load the shell and get it's opts, etc.

**Comment by shahzebsiddiqui** [comment]
Okay i will try to implement ``Shell`` class tomorrow.
A few questions 
1. Any preference where the ``Shell`` class should be declared with respect to modules. Should this a new file let's say ``buildtest/utils/shell.py``, or part of ``buildtest/buildsystem/base.py`` or somewhere else?
2. I don't think ``self.shell.shebang`` should be available in Shell class based on https://github.com/buildtesters/buildtest/pull/290#discussion_r416881599. The shell script with a shebang ``#!/bin/bash`` will run with the ``/bin/bash`` as the shell interpreter. The script can also be run directly but when we need to pass shell options then we need to do something like ``bash [option] script.sh``. I think we need to make a distinction between ``shell`` and ``shebang`` key. The shebang key is used to tweak the test script shebang line whereas the shell key  is used to tweak how to run the test. The more i think about shell key this looks more suitable for LocalExecutor behavior.

**Comment by vsoch** [comment]
> Any preference where the Shell class should be declared with respect to modules. Should this a new file let's say buildtest/utils/shell.py, or part of buildtest/buildsystem/base.py or somewhere else?

Yes! I would even go as far to say that (this is cool enough) that someone might want to use the class as a helper library even outside of build test (meaning it could exist as it's own Python module) but since this is very specific to buildtest, unlike schemas, let's keep it inside the buildtest module. I think we should put it where it intuitively would be found, so akin to a BuildTestCommand, I would do Shell under:

```
buildtest/utils/shell.py
```

so we would do:

```python
from buildtest.utils.shell import Shell
```

I don't think there is anything special about the shell, but if you think it's more consitent you could also do 

```python
from buildtest.utils.shell import BuildTestShell
```
Even though I'm not sure there is a thing called a "build test shell." I'll leave that up to you.

> I don't think self.shell.shebang should be available in Shell class based on #290 (comment). The shell script with a shebang #!/bin/bash will run with the /bin/bash as the shell interpreter. The script can also be run directly but when we need to pass shell options then we need to do something like bash [option] script.sh. I think we need to make a distinction between shell and shebang key. The shebang key is used to tweak the test script shebang line whereas the shell key is used to tweak how to run the test. The more i think about shell key this looks more suitable for LocalExecutor behavior.

Sure, that works. If you want to have a function that will generate a shebang based on a shell, that could be provided, and perhaps used if the user hasn't set anything. Again, totally up to you, and I'm happy to review what you decide.

File: buildtest/utils/shell.py
Comment (vsoch) [inline buildtest/utils/shell.py:5]: Perhaps a default to bash?

File: buildtest/utils/shell.py
Comment (vsoch) [inline buildtest/utils/shell.py:25]: The function opts would immediately override the class variable of the same name.

File: buildtest/utils/shell.py
Comment (vsoch) [inline buildtest/utils/shell.py:74]: These should be done in init.

**Comment by shahzebsiddiqui** [comment]
Hey @vsoch,

So i have created the Shell class which can be invoked directly so i will share how this works before i integrate into Builderbase

This is a simple call to invoke a bash shell and use the get method to retrieve all the shell attributes
```
>>> shell = Shell("bash")
>>> shell.get()
{'name': 'bash', 'opts': '', 'path': '/usr/bin/bash', 'shebang': '#!/usr/bin/bash'}
```

We can retrieve the shell path by invoking 

```
>>> shell.path
'/usr/bin/bash'
```

The class automatically updates shell path if the ``name`` attribute is changed and if i understood correctly you wanted to use decorators for this purpose

Note i change ``shell.name="sh"`` and ``shell.path ``and ``shell.shebang`` also changed
```
>>> shell.name="sh"
>>> shell.get()
{'name': 'sh', 'opts': '', 'path': '/usr/bin/sh', 'shebang': '#!/usr/bin/sh'}

```

We can also update the opts which would be useful during a shell change 

```
>>> shell.opts = "-c"
>>> shell.get()
{'name': 'sh', 'opts': '-c', 'path': '/usr/bin/sh', 'shebang': '#!/usr/bin/sh'}
```

If user specify invalid shell we raise an exception such as this example

```
>>> shell.name = "bash1"
>>> shell.get()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/mxg-hpc/users/ssi29/buildtest/buildtest/utils/shell.py", line 65, in get
    self.shell["path"] = self.path
  File "/mxg-hpc/users/ssi29/buildtest/buildtest/utils/shell.py", line 53, in path
    raise BuildTestError(f"Can't find program: {self.name}")
buildtest.exceptions.BuildTestError: "Can't find program: bash1"
```

I assumed the input in Shell constructor must be a string so i raise an exception if it's not enforced.

File: buildtest/utils/shell.py
Comment (vsoch) [inline buildtest/utils/shell.py:84]: Test this out - most objects already can be serialized as dict so it’s about setting properties.

**Review by vsoch — CHANGES_REQUESTED** [review]

File: buildtest/utils/shell.py
Comment (shahzebsiddiqui) [reply buildtest/utils/shell.py:5]: i can do that so make argument ``shell`` as optional argument so that we can invoke the class as
`` shell = Shell()`` which defaults to bash.

File: buildtest/utils/shell.py
Comment (shahzebsiddiqui) [reply buildtest/utils/shell.py:25]: so what are you suggesting we do? Remove ``opts`` method?

File: buildtest/utils/shell.py
Comment (shahzebsiddiqui) [reply buildtest/utils/shell.py:5]: So how is this

```
>>> shell = Shell()
>>> shell.get()
{'name': 'bash', 'opts': '', 'path': '/usr/bin/bash', 'shebang': '#!/usr/bin/bash'}
>>> sh = Shell("sh")
>>> sh.get()
{'name': 'sh', 'opts': '', 'path': '/usr/bin/sh', 'shebang': '#!/usr/bin/sh'}

```

File: buildtest/utils/shell.py
Comment (shahzebsiddiqui) [reply buildtest/utils/shell.py:74]: ok i have done this in init also, but found that i also need this check in ``path`` method when there is a change in name attribute meaning ``shell.name = "python1"`` we report an error otherwise the path and shebang will be ``None``

File: buildtest/utils/shell.py
Comment (vsoch) [reply buildtest/utils/shell.py:25]: It would be appropriate to parse original opts in the init, but then have this function more properly named based on what it does, e.g., shell.set_opts()

File: buildtest/utils/shell.py
Comment (vsoch) [reply buildtest/utils/shell.py:74]: Just create a function to handle these operations and call it in init. If the user of the Shell class wants to call it again after, they are free to do that.

**Comment by shahzebsiddiqui** [comment]
I am having hard time understanding based on your comments on what you want. This would be where you can issue a PR to my branch with how you plan to use the Shell class. I guess once this class is done i will update BuilderBase to use the Shell class.

**Comment by vsoch** [comment]
All set. Take a look!

**Comment by shahzebsiddiqui** [comment]
It's not perfect but i tested it out and didn't intend user to update the shell path which will update the name since that is quite hard to do

For example in this example the name get's same as path.
```
>>> shell.path="/usr/bin/python"
>>> shell.get()
{'name': '/usr/bin/python', 'opts': '', 'path': '/usr/bin/python', 'shebang': '#!/usr/bin/python'}
```

Another thing is if user changes the name, the path and shebang are not updated automatically as shown below

```
>>> shell.name="sh"
>>> shell.get()
{'name': 'sh', 'opts': '', 'path': '/usr/bin/python', 'shebang': '#!/usr/bin/python'}
```

At the moment, i don't really care as buildtest will invoke Shell once and get the appropriate shell. I kinda liked my implementation before, where path and shebang were only changed based on name change.

**Comment by vsoch** [comment]
I can only implement what you’ve clearly communicated as being important. Please make a list of billets for how you want it to work and I’ll update the class.

**Comment by shahzebsiddiqui** [comment]
Features for Shell class
- [x] User can invoke ``Shell()`` which should default to bash 
- [x] User can update the class attribute ``opts`` so something like ``shell.opts="-v"`` this may be useful to override options. 
- [x] Catch exception when invalid path detected during init i.e ``shell = Shell("python4")``
- [] Catch exception when invalid path is result of change in attribute ``name`` (i.e self.name="python4")
```
>>> shell.name="python4"
>>> shell.get()
{'name': 'python4', 'opts': '', 'path': '/usr/bin/bash', 'shebang': '#!/usr/bin/bash'}
```
- [] User can update class attribute ``name`` let's assume shell is an instance of Shell we can do ``shell.name="python"`` which will update the path and shebang automatically (not working now)

```
>>> shell = Shell()
>>> shell.get()
{'name': 'bash', 'opts': '', 'path': '/usr/bin/bash', 'shebang': '#!/usr/bin/bash'}
>>> shell.name="sh"
>>> shell.get()
{'name': 'sh', 'opts': '', 'path': '/usr/bin/bash', 'shebang': '#!/usr/bin/bash'}

```

- [] Remove the feature where the class attribute path can be changed once object is called so remove ``@path.setter`` https://github.com/shahzebsiddiqui/buildtest-framework/blob/update_shell_logic/buildtest/utils/shell.py#L66
This results in weird things where user can do something really bizarre

```
>>> shell.path="hostname"
>>> shell.get()
{'name': 'hostname', 'opts': '', 'path': '/usr/bin/hostname', 'shebang': '#!/usr/bin/hostname'}
```

**Comment by vsoch** [comment]
@shahzebsiddiqui you can't have path update name and vice versa - that leads to an infinite loop :)

**Comment by vsoch** [comment]
The way that shell is designed now works as expected to instantiate the Shell, and I don't see a use case being that a user would need to go in and start customizing random things. So I'd suggest that you keep as is and open these issues when they are actually proven to be issues.

**Comment by shahzebsiddiqui** [comment]
> The way that shell is designed now works as expected to instantiate the Shell, and I don't see a use case being that a user would need to go in and start customizing random things. So I'd suggest that you keep as is and open these issues when they are actually proven to be issues.

Agreed, right now i don't have a use-case of using Shell class directly I think this was suggestion you made. Anyhow, i integrate this into BuilderBase and is ready for final review.

File: buildtest/buildsystem/base.py
Comment (vsoch) [inline buildtest/buildsystem/base.py]: These 5 lines can just be

```python
shell = Shell(self.recipe.get("shell", "bash"))
```

File: buildtest/buildsystem/base.py
Comment (vsoch) [inline buildtest/buildsystem/base.py]: The reason to have the self.shell as the shell class is that we would do:

```
self.shell = Shell()
```
and then refer to `self.shell.shebang` and not need to do shell.get() (this doesn't make sense to me). If the shebang can be included with the shell init and then referenced in the same way, this would also be cleaner code.

File: buildtest/buildsystem/base.py
Comment (vsoch) [inline buildtest/buildsystem/base.py]: This would be `self.shell.name`

File: buildtest/buildsystem/base.py
Comment (vsoch) [inline buildtest/buildsystem/base.py]: If you do self.recipe.get("env", []) you don't need to do `if pairs` but can just loop through them still (an empty list).

File: buildtest/buildsystem/base.py
Comment (vsoch) [inline buildtest/buildsystem/base.py]: Why is this added to metadata?

File: buildtest/buildsystem/base.py
Comment (vsoch) [inline buildtest/buildsystem/base.py]: This again should reference properties and not dictionary items, given that you are interacting with the class.

File: buildtest/utils/command.py
Comment (vsoch) [inline buildtest/utils/command.py]: Why are these changes showing up here when I already made them in another PR? I wrote these lines...

**Review by vsoch — CHANGES_REQUESTED** [review]

File: buildtest/utils/command.py
Comment (shahzebsiddiqui) [reply buildtest/utils/command.py]: I was testing something, this was an accident my bad ill undo this, ill copy the file from devel since i didnt make any change to the file.

File: buildtest/buildsystem/base.py
Comment (shahzebsiddiqui) [reply buildtest/buildsystem/base.py]: Ill remove it for now, it was not clear how this metadata will be used going forward one thought was to use the metadata to write to a json file for reporting (``buildtest build report``) #284

File: buildtest/buildsystem/base.py
Comment (shahzebsiddiqui) [reply buildtest/buildsystem/base.py]: This is now fixed along with the other stuff mentioned, feel free to resolve what you think is addressed.

File: buildtest/buildsystem/base.py
Comment (vsoch) [inline buildtest/buildsystem/base.py:272]: This whole section is much cleaner than it was before - good job!

File: buildtest/buildsystem/base.py
Comment (vsoch) [inline buildtest/buildsystem/base.py]: This is commented out and should not be there.

File: buildtest/utils/command.py
Comment (vsoch) [inline buildtest/utils/command.py]: This was written by me, and shouldn't appear like this. It suggests that you didn't pull the merged commits from my contribution. It's not really okay to take credit for my work, so this needs to be fixed.

**Comment by shahzebsiddiqui** [comment]
Travis is failing because ``shell`` is part of ``metadata`` in the test https://github.com/shahzebsiddiqui/buildtest-framework/blob/update_shell_logic/tests/buildsystem/test_base.py#L54 which i dont understand exactly why this was being done in the test. I am guessing we are checking if each env variable is present.

 since i removed ``shell`` it i can put it back in but it would be ``self.metadata["shell"]=self.shell.name`` Originally it was ``self.metadata["shell"]=self.get_shell()`` which was returning the name of the shell

File: buildtest/buildsystem/base.py
Comment (shahzebsiddiqui) [reply buildtest/buildsystem/base.py]: See https://github.com/buildtesters/buildtest/pull/290#commitcomment-38882276. I introduced this bug by commenting this out. I guess we still need the condition ``if pairs`` when a Buildspec doesn't define ``env`` section we just skip the whole thing

**Comment by shahzebsiddiqui** [comment]
I have added ``shell`` back into ``self.metadata`` i didn't realize this was used in https://github.com/shahzebsiddiqui/buildtest-framework/blob/update_shell_logic/tests/buildsystem/test_base.py#L54 i don't really like the way this was setup as ``self.metadata["shell"]`` is not used anywhere in the code except the line i mentioned in test_base.py
The test should be passing, if you dont like this i would suggest to remove the check in ``test_base.py``

**Comment by vsoch** [comment]
Please address why you have the commits from my Capturing class, already commit, that are included in this PR as your own. This absolutely cannot be merged in that state, as you would be stealing my work.

File: buildtest/utils/command.py
Comment (shahzebsiddiqui) [reply buildtest/utils/command.py]: I am not sure what happened i do see your git commit **Fix/command out err (#289)** in my commit history https://github.com/shahzebsiddiqui/buildtest-framework/commits/update_shell_logic 
The only other explanation could be that when i copy-paste to fix this i applied black so it's showing a diff.
I am bit afraid to do a merge from devel now

**Comment by shahzebsiddiqui** [comment]
> Please address why you have the commits from my Capturing class, already commit, that are included in this PR as your own. This absolutely cannot be merged in that state, as you would be stealing my work.

I am trying to figure this out see https://github.com/buildtesters/buildtest/pull/290#discussion_r418695856

**Comment by vsoch** [comment]
If you have integrated devel and this persists, then you must have done something (copy paste??) to have messed that up. Please start fresh with devel, and then cherry pick the commits from here, not including the one that re-does the work that I already did. I don't know what you did with git, but it's not including my work, but counting it here as new work (authored by you) which is really not ok :(

**Comment by shahzebsiddiqui** [comment]
Let me try to merge from devel and see if it fixes the problem i have done it locally now checking if there were any merge conflicts before i push.

**Comment by vsoch** [comment]
Just don't merge this PR, whatever you do.

**Comment by vsoch** [comment]
Ah so that was the issue? You had copy pasted my file without merging (rebasing with devel?) Just curious... why?

**Comment by shahzebsiddiqui** [comment]
Okay i think the merge https://github.com/shahzebsiddiqui/buildtest-framework/commit/4a1b7168092f4db2928d6156388d84a046cac6c9  fixed the commit diff for ``buildtest/utils/command.py``  so now you don't see all the changes i made due to copy-paste sorry about that. I have no clue how this happened.

File: buildtest/buildsystem/base.py
Comment (shahzebsiddiqui) [reply buildtest/buildsystem/base.py]: This is fixed now

File: buildtest/buildsystem/base.py
Comment (vsoch) [inline buildtest/buildsystem/base.py:503]: If the cmd is created here and then has lines added to it, why do we need to create it as an empty list first? E.g., just do:

```python
cmd = [self.shell.path, self.shell.opts, testfile]
```

Comment (shahzebsiddiqui) [reply buildtest/buildsystem/base.py:503]: This is fixed now in https://github.com/shahzebsiddiqui/buildtest-framework/commit/b18e1c53fbb42e081b6e016daa911d63677a9295

**Comment by shahzebsiddiqui** [comment]
@vsoch i have added regression test for Shell class in ``test/utils/test_shell.py``  in https://github.com/shahzebsiddiqui/buildtest-framework/commit/b18e1c53fbb42e081b6e016daa911d63677a9295

Good news coverage is 100% for the initial implementation.

**Review by vsoch — APPROVED** [review]

**Pull request merged**
